packages/teleterm/src/ui/DocumentCluster/ClusterResources/useServerSideResources.ts

```diff
@@ -0,0 +1,153 @@
+  const [keys, setKeys] = useState<string[]>([]);
+  const [agentFilter, setAgentFilter] = useState<AgentFilter>({
+    sort: getDefaultSort(),
+  });
+  const [pageCount, setPageCount] = useState({
+    to: 0,
+    from: 0,
+    total: 0,
+  });
+
+  const limit = 15;
+
+  function getFetchCallback(params: ServerSideParams) {
+    switch (resource) {
+      case ResourceKind.Server:
+        return retryWithRelogin(ctx, documentUri, clusterUri, () =>
+          ctx.clustersService.fetchServers(params)
+        );
+      default: {
+        throw new Error(`Fetch not implemented for: ${resource}`);
+      }
+    }
+  }
+
+  // startKey is used here as a way to paginate through agents returned from
+  // their respective rpcs.
+  const [fetchAttempt, fetch] = useAsync(async (startKey: string) => {
+    return getFetchCallback({
+      ...agentFilter,
+      limit,
+      clusterUri,
+      startKey,
+    });
+  });
+
+  // If there is no startKey at the current page's index, there is no more data to get
+  function nextPage() {
+    const proposedKey = keys[pageIndex];
+    if (proposedKey) {
+      return () => {
+        fetch(proposedKey);
+        setPageIndex(pageIndex + 1);
+      };
+    }
+    return null;
+  }
+
+  // If we are at the first page (index 0), we cannot fetch more previous data
+  function prevPage() {
+    const newPageIndex = pageIndex - 1;
+    if (newPageIndex > -1) {
+      return () => {
+        // if we got pageIndex - 1, we'd get the same data as the previous "next" we just called.
+        // we have to go back 2. which is why we use newPageIndex instead
+
+        // example after going from pageIndex = 0 to pageIndex = 1:
+        // ['startKeyFromInitialFetch', 'goEvenFurtherStartKey']
+        //     ^- just came from here          ^- we are here
+        //
+        // so in order to get the start key for the page BEFORE this one, we go back twice.
+        // this could also be written/read as keys[pageIndex - 2]
+        fetch(keys[newPageIndex - 1]);
+        setPageIndex(newPageIndex);
+      };
+    }
+    return null;
+  }
+
+  useEffect(() => {
+    // when agent filter changes we consider it a 'new' search so we should pass an empty
+    // startKey to get the beginning of the list with new filters, and set pageIndex back to zero
+    fetch('');
+    setPageIndex(0);
+  }, [agentFilter]);
+
+  // when we receive data from fetch, we store the startKey (or lack of) according to the current
+  // page index. think of this as "this page's nextKey".
+  // "why don't we just name it nextKey then?"
+  // Mostly because it's called startKey almost everywhere else in the UI, and also because we'd have the same issue
+  // for prevPage if we swapped named, and this comment would be explaining "this page's startKey".
+  useEffect(() => {
+    const newKeys = [...keys];
+    newKeys[pageIndex] = fetchAttempt.data?.startKey;
+    setKeys(newKeys);
+
+    setPageCount(getPageCount());
+  }, [fetchAttempt.data]);
```

Comment:
This is where I'd imagine we would put something like `addDataIfMissing` type of call to the clustersService cache for the current resource